### PR TITLE
AGDROID-683 - Use Aerogear Auth SDK - Replace AppAuth SDK/Auth Services

### PR DESCRIPTION
**Pull request by tomjackman**
### Motivation
PR to replace the hard AppAuth dependency with the Aerogear Android IDM SDK. - https://issues.jboss.org/browse/AGDROID-683

### Aerogear IDM SDK Implementation Progress
- [x] Add mobile services JSON file. https://github.com/feedhenry/mobile-security-android-template/pull/36/commits/b67be2323fe2609a89c00c4de11d3706f2238fef
- [x] Add build.gradle config https://github.com/feedhenry/mobile-security-android-template/pull/36/commits/4d07b824c198f1c582a40a8efb62e41365da87cd
- [x] Expose a single instance of the Mobile Core with dagger. https://github.com/feedhenry/mobile-security-android-template/pull/36/commits/d62f75f5c954b352059280483dea6cffa9828606
- [x] Expose a single instance of the Auth Service with dagger. https://github.com/feedhenry/mobile-security-android-template/pull/36/commits/afb7d55863a598516dccb93099441fa560a3337e
- [x] Add intent listener in activity.
- [x] Allow auth requests to the system browser. https://github.com/feedhenry/mobile-security-android-template/pull/36/commits/4c2753bc737d05a94b53e89e8428cc26f339a22b
- [x] Implement getRoles https://github.com/feedhenry/mobile-security-android-template/pull/36/commits/c206222aa567f0318514a21287102a6278ce98a9
- [x] Implement getAccessToken https://github.com/feedhenry/mobile-security-android-template/pull/36/commits/90a3dae7f842b50b7a1c4b96c6998e579e6a6ee2
- [x] Update auth routing flow https://github.com/feedhenry/mobile-security-android-template/pull/36/commits/508be3ca348cee43036d61c46a6a34a3c12f1787
- [x] Implement logout https://github.com/feedhenry/mobile-security-android-template/pull/36/commits/508be3ca348cee43036d61c46a6a34a3c12f1787
- [x] Update Access Control Functionality https://github.com/feedhenry/mobile-security-android-template/pull/36/commits/c206222aa567f0318514a21287102a6278ce98a9
- [x] Confirm Cert Pinning is Working on the Auth Screen https://github.com/feedhenry/mobile-security-android-template/pull/36/commits/a83a09d9baaae5fc83a66782680a94d71c451d77
- [x] Remove AppAuth dependency. https://github.com/feedhenry/mobile-security-android-template/pull/36/commits/f1b9868dffe2293d0221522fc03564adaa442683
- [x] Remove Auth Config Class https://github.com/feedhenry/mobile-security-android-template/pull/36/commits/bd72630f5bb345b2bb951ddffbd55eb736c63514
- [x] Remove old keycloak auth config https://github.com/feedhenry/mobile-security-android-template/pull/36/commits/14b2ca75e3a2e9c340c843d254ef665c544dca60
- [x] Remove Auth State Service Class https://github.com/feedhenry/mobile-security-android-template/pull/36/commits/a83a09d9baaae5fc83a66782680a94d71c451d77
- [x] Remove unnecessary imports https://github.com/feedhenry/mobile-security-android-template/pull/36/commits/7cc5a04bae76304049f1c1e90cb8a767ce91c805
- [x] Update Tests https://github.com/feedhenry/mobile-security-android-template/pull/36/commits/457b9cc4368046798ff50146ab143bae22213387
- [x] Update Logout to use callbacks https://github.com/feedhenry/mobile-security-android-template/pull/36/commits/da2f210dc889258e24bf54789c4262c1b3bf5b66
- [x] Re-write of all auth documentation.  https://github.com/feedhenry/mobile-security/pull/66
- [x] Re-write of access control docs https://github.com/feedhenry/mobile-security/pull/66
- [x] Re-write of cert pinning/authenticated network calls docs (where necessary) https://github.com/feedhenry/mobile-security/pull/66
- [x] Update main docs to point to new Keycloak config. https://github.com/feedhenry/mobile-security-android-template/pull/36/commits/53a57f7a0fb7ea1ca2129b398bf73b588eae2971

**Note**: This PR needs to be merged at the same time as https://github.com/feedhenry/mobile-security/pull/66 due to the hosted docs pointing live at the code in this repo.

### User Feature Progress
- [x] A user can click the login button and login through the system browser.
- [x] A user should see their profile after logging in.
- [x] A user should be able to see their client and realm roles.
- [x] A user should be able to logout.
- [x] A user should be able to send authenticated requests to protected resources.
- [x] A user should be able to use access control functionality.

### Notes
The Callback interface has been renamed to CallbackHandler since this clashes with the Aerogear SDK Callback class name.

### CODE REVIEW

File: app/src/main/java/com/feedhenry/securenativeandroidtemplate/domain/services/MobileCoreService.java [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+
+import javax.inject.Inject;
+import javax.inject.Singleton;
+
+/**
+ * Created by tjackman on 22/02/18.
+ */
+
+@Singleton
+public class MobileCoreService {
```
Comment (wei-lee) [inline app/src/main/java/com/feedhenry/securenativeandroidtemplate/domain/services/MobileCoreService.java]: is this actually required? can we not just return an instance of MobileCore in the `SecureApplicationModule.java` class?

File: app/src/main/java/com/feedhenry/securenativeandroidtemplate/features/authentication/providers/KeycloakAuthenticateProviderImpl.java [hunk 2]
```diff
@@ -55,61 +63,54 @@
...
     @Inject
     Context context;
 
     @Inject
-    public KeycloakAuthenticateProviderImpl(@NonNull Context context, AppConfiguration appConfiguration, AuthStateService authStateService) {
+    public KeycloakAuthenticateProviderImpl(@NonNull Context context, AppConfiguration appConfiguration, AuthStateService authStateService, MobileCoreService mobileCoreService) {
         this.context = context;
         this.authenticationConfiguration = appConfiguration.getAuthConfiguration();
         this.authStateService = authStateService;
+        this.mobileCoreService = mobileCoreService;
```
Comment (wei-lee) [inline app/src/main/java/com/feedhenry/securenativeandroidtemplate/features/authentication/providers/KeycloakAuthenticateProviderImpl.java]: should be able to use DI to inject the AuthService directly. Do we actually need the MobileCoreService here?

File: app/src/main/java/com/feedhenry/securenativeandroidtemplate/domain/services/MobileCoreService.java (hunk 1 above)
Comment (tomjackman) [reply app/src/main/java/com/feedhenry/securenativeandroidtemplate/domain/services/MobileCoreService.java]: @wei-lee Sorry, missed a commit from mondays work. I have brought it in now. Can you re-check?

File: app/src/main/java/com/feedhenry/securenativeandroidtemplate/features/authentication/providers/KeycloakAuthenticateProviderImpl.java [hunk 3]
```diff
@@ -55,61 +62,55 @@
...
         this.authenticationConfiguration = appConfiguration.getAuthConfiguration();
         this.authStateService = authStateService;
+
+        AuthService authService = mobileCore.getInstance(AuthService.class);
+        AuthServiceConfiguration authServiceConfig = new AuthServiceConfiguration.AuthConfigurationBuilder()
+                .withRedirectUri("com.feedhenry.securenativeandroidtemplate:/callback")
+                .build();
+
+        // Only invoke this once, every subsequent retrieval of the AuthService will retrieve the same, already initialized, instance.
+        authService.init(context, authServiceConfig);
```
Comment (wei-lee) [inline app/src/main/java/com/feedhenry/securenativeandroidtemplate/features/authentication/providers/KeycloakAuthenticateProviderImpl.java]: I think we can move it to use DI as well

File: app/build.gradle [hunk 4]
```diff
@@ -1,12 +1,12 @@
 apply plugin: 'com.android.application'
 
 android {
-    compileSdkVersion 25
-    buildToolsVersion '25.0.3'
+    compileSdkVersion 27
+    buildToolsVersion '26.0.2'
```
Comment (secondsun) [inline app/build.gradle]: current build tools is 27.0.3

File: app/build.gradle [hunk 5]
```diff
@@ -44,14 +50,21 @@ android {
...
-    compile 'com.android.support:support-v4:25.3.1'
-    compile 'net.openid:appauth:0.7.0'
-    compile 'com.android.support:cardview-v7:25.3.1'
+    compile "com.android.support:appcompat-v7:$android_support_version"
+    compile "com.android.support:animated-vector-drawable:$android_support_version"
+    compile "com.android.support:customtabs:$android_support_version"
+    compile "com.android.support:design:$android_support_version"
+    compile "com.android.support:cardview-v7:$android_support_version"
+    compile "com.android.support:support-v4:$android_support_version"
+    compile "com.android.support.constraint:constraint-layout:1.0.2"
```
Comment (secondsun) [inline app/build.gradle]: compile is being deprecated, please use implementation instead

**Comment by secondsun** [comment]
Looking great.  I had a small problem with using self signed certificates on Android 22 that we can work out.  Additionally the sdk has been released to maven central so we can use 0.1.0-2018-07 and not need a local build of the SDK.

@danielpassos and I are working on doing weekly releases of whatever is in github.

**Comment by danielpassos** [comment]
@secondsun The first one is already there => http://search.maven.org/#artifactdetails%7Corg.aerogear%7Ccore%7C0.1.0-2018-07%7Caar

**Comment by tomjackman** [comment]
1. I will need to wait until https://github.com/aerogear/aerogear-android-sdk/pull/118 is merged and is published in a new version of the SDK before merging this PR so people don't need to build the SDK locally. 

2. There is an issue with the Circle CI builds with this update. The `.gradle` folder is being cached currently [here](https://github.com/feedhenry/mobile-security-android-template/blob/master/circle.yml#L24). Going to remove this line, run a build, then add it back into to stop the build failing due to these new changes.

**Comment by secondsun** [comment]
@TommyJ1994  A patch to test the prerelease of the android SDK

From 859c16c10cdea3ac7b01a6c7b757ed48c8b9909b Mon Sep 17 00:00:00 2001
From: Summers Pittman <secondsun@gmail.com>
Date: Tue, 6 Mar 2018 15:14:56 -0500
Subject: [PATCH] Using prereleased android

---
 app/build.gradle | 4 ++--
 build.gradle     | 1 +
 2 files changed, 3 insertions(+), 2 deletions(-)

diff --git a/app/build.gradle b/app/build.gradle
index bce1241..20d33bd 100644
--- a/app/build.gradle
+++ b/app/build.gradle
@@ -63,8 +63,8 @@ dependencies {
     implementation "com.android.support:cardview-v7:$android_support_version"
     implementation "com.android.support:support-v4:$android_support_version"
     implementation "com.android.support.constraint:constraint-layout:1.0.2"
-    implementation 'org.aerogear:core:0.1.0-SNAPSHOT'
-    implementation 'org.aerogear:auth:0.1.0-SNAPSHOT'
+    implementation 'org.aerogear:core:0.1.0-2018-09'
+    implementation 'org.aerogear:auth:0.1.0-2018-09'
     implementation 'com.squareup.okhttp3:okhttp:3.9.0'
     implementation 'com.datatheorem.android.trustkit:trustkit:1.0.1@aar'
     implementation 'com.scottyab:rootbeer-lib:0.0.6'
diff --git a/build.gradle b/build.gradle
index cdf02a4..f585018 100644
--- a/build.gradle
+++ b/build.gradle
@@ -30,6 +30,7 @@ allprojects {
         mavenLocal()
         jcenter()
         google()
+        maven() {url 'https://oss.sonatype.org/content/repositories/orgaerogear-1018/'}
     }
 }
 
-- 
2.14.3

**Comment by tomjackman** [comment]
@secondsun Tested with the prerelease version and works fine. Is the `0.1.0-2018-09` version going to remain the same for the release? I can bump the version now if that's so.

**Comment by secondsun** [comment]
@TommyJ1994 Yes.  I just told passos to release to central :)

**Comment by tomjackman** [comment]
@secondsun Version bumped now. Functionality/Tests are fine. Should be good to go.

The build still might fail because of the outdated cached `.gradle` folder in Circle CI, but I can fix this in another PR.

**Comment by tomjackman** [comment]
@secondsun So the failing test is the `NoteCrudlServiceTest`. This isn't related to the auth changes. 

This passes fine locally, but the only error from the Circle CI logs is:

```
com.feedhenry.securenativeandroidtemplate.domain.services.NoteCrudlServiceTest > testService FAILED
    java.lang.UnsupportedOperationException
```

I hit this issues before and can be fixed by deleting the `.gradle` folder inside the application. I did this as part of https://github.com/feedhenry/mobile-security-android-template/pull/36/commits/d3e3d94201b97a714f7184a9856b8452fe432647 but no use.

**Comment by tomjackman** [comment]
`java.lang.UnsupportedOperationException: Robolectric does not support API level 27.`

Looks like its caused by that `@config` in the test.

**Comment by secondsun** [comment]
Weird.  I just checked the source and robolectric supports 27

https://github.com/robolectric/robolectric/blob/master/robolectric/src/main/java/org/robolectric/internal/SdkConfig.java#L26

**Comment by tomjackman** [comment]
Yeah, I checked it too and was wondering why. Emulating using 26 works fine, but hit another unrelated error: `android.content.res.Resources$NotFoundException: com.feedhenry.securenativeandroidtemplate:xml/network_security_config`. This is supported in API level 26 though. Hmm.

**Comment by secondsun** [comment]
Emulators in CI are painful.  Maybe we should just disable the test and flag fixing the issue later?

**Comment by tomjackman** [comment]
@secondsun I have merged it for now since the test is failing because of non-auth related change. Issue created to fix in the future: https://github.com/feedhenry/mobile-security-android-template/issues/39

**Pull request merged**